monitoring/ping.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"context"
+	"log"
+
+	"github.com/gravitational/satellite/agent"
+	"github.com/gravitational/satellite/agent/health"
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+	"github.com/gravitational/trace"
+	serf "github.com/hashicorp/serf/client"
+	ping "github.com/sparrc/go-ping"
+)
+
+const (
+	PingCheckerID = "ping-checker"
+)
+
+func NewPingChecker(serfRPCaddr string, role string) health.Checker {
+	return &PingChecker{
+		serfRPCAddr: serfRPCaddr,
+		role:        role,
+	}
+}
+
+// PingChecker is a checker that verifies if ping between Master Nodes and other
+// ones is below a specified threshold
+type PingChecker struct {
+	serfRPCAddr string
+	role        string
+}
+
+// Name returns the check name
+// Implements health.Checker
+func (c *PingChecker) Name() string {
+	return PingCheckerID
+}
+
+// Check verifies that all nodes' ping with Master Nodes is lower than the
+// desired threshold
+// Implements health.Checker
+func (c *PingChecker) Check(ctx context.Context, r health.Reporter) {
+	RttThreshold := 100     // ms
+	slidingWindowSize := 10 // number of ping results to consider per iteration
+
+	// set Probe to be running
+	r.Add(&pb.Probe{
+		Checker: c.Name(),
+		Status:  pb.Probe_Running,
+	})
+
+	// fetch serf config and intantiate client
+	clientConfig := serf.Config{
+		Addr: c.serfRPCAddr,
+	}
+	client, err := serf.ClientFromConfig(clientConfig)
+	if err != nil {
+		return // nil, trace.Wrap(err, "failed to connect to serf")
+	}
+	defer client.Close()
+
+	// only run PingCheck between masters
+	role := c.role
+	if role != agent.RoleMaster {
+		r.Add(&pb.Probe{
+			Checker: c.Name(),
+			Status:  pb.Probe_Terminated,
+		})
+		return // skip check if the node is not a master
+	}
+
+	// ping other Master nodes and store results in Serf
+	nodes := client.Members()
+	for node := range nodes {
+		// FIXME: BEGINIF if other node is master {
+		pinger, err := ping.NewPinger(node.Addr)
```

Comment:
This library sends ICMP packets as far as I can see, just like regular ping. Are you sure we can rely on ICMP not being blocked b/w the nodes? I think this is often the case.

Reply:
Yes, that's a normal ICMP ping so I guess that could be true in some ENVs.

Well, the two other alternatives I evaluated were:
1- create a listener on a custom port and send/receive messages on that instead, but we'll have to ask for an additional port which is even worse probably, as opposed to allow ICMP internally
2- I guess it could be possible to send/receive messages from within the checker itself sending a fixed size packet but that would essentially mean reimplementing a "simpler ping"... right?

Reply:
You don't necessarily need an additional port, agent already serves API on port 7575 so you could just use that.

Also, I wonder if serf itself already has some built-in mechanism to determining latency b/w its nodes? I would assume it does, with it being a cluster membership tool. I suggest you research it, e.g. take a look at https://www.serf.io/docs/commands/rtt.html and see if we can use it instead.

Reply:
Seems that Serf RTT is a really crude implementation done in CLI and returning strings

https://github.com/hashicorp/serf/blob/master/cmd/serf/command/rtt.go

doesn't seem like a good choice to me :man_shrugging:

Reply:
Well, I wasn't proposing to call a CLI command obviously. It uses serf API, doesn't it?

Reply:
it's actually just creating an RPCClient and then use:

https://github.com/hashicorp/serf/blob/master/client/rpc_client.go#L370

on node1 and node2, then calling:

https://github.com/hashicorp/serf/blob/master/coordinate/coordinate.go#L119

between the two. I guess I could use the same approach and call those APIs myself, just not the RTT tool directly.

Reply:
I tried and failed implementing this, so for now I guess we'll have all the rest set and then I'll wrap it up by adding this one last piece.

Reply:
@eldios What's the issue? Serf "rtt" command that you linked to the source of above does exactly what we need, no? So you can just implement our check same way?

Reply:
Managed to implement it.